cpp/src/binning/bin.cu

```diff
@@ -0,0 +1,216 @@
+
+template <typename T, typename LeftComparator, typename RightComparator>
+struct bin_finder {
+  bin_finder(detail::device_span<T> left_span, detail::device_span<T> right_span)
+    : m_left_span(left_span), m_right_span(right_span)
+  {
+  }
+
+  __device__ size_type operator()(thrust::pair<T, bool> input_value) const
+  {
+    // Immediately return sentinel for null inputs.
+    if (!input_value.second) return NULL_VALUE;
+
+    T value = input_value.first;
+    auto bound =
+      thrust::lower_bound(thrust::seq, m_left_span.begin(), m_left_span.end(), value, m_left_comp);
+
+    // Exit early and return sentinel for values that lie below the interval.
+    if (bound == m_left_span.begin()) { return NULL_VALUE; }
+
+    // We must subtract 1 because lower bound returns the first index
+    // _greater than_ the value. This is safe because bound == m_left edges
+    // would already have triggered a NULL_VALUE return above, so there's
+    // no risk of negative values or wraparound for -1 here.
+    auto index = bound - m_left_span.begin() - 1;
+    return (m_right_comp(value, m_right_span[index])) ? index : NULL_VALUE;
+  }
+
+  detail::device_span<T> m_left_span{};   // The range of data containing all left bin edges.
+  detail::device_span<T> m_right_span{};  // The range of data containing all right bin edges.
+  LeftComparator m_left_comp{};           // Comparator used for left edges.
+  RightComparator m_right_comp{};         // Comparator used for left edges.
+};
+
+// Functor to identify rows that should be filtered out based on the sentinel set by
+// bin_finder::operator().
+struct filter_null_sentinel {
+  __device__ bool operator()(size_type i) { return i != NULL_VALUE; }
+};
+
+/// Bin the input by the edges in left_edges and right_edges.
+template <typename T, typename LeftComparator, typename RightComparator, bool InputIsNullable>
+std::unique_ptr<column> bin(column_view const& input,
+                            column_view const& left_edges,
+                            column_view const& right_edges,
+                            rmm::mr::device_memory_resource* mr)
+{
+  auto output = cudf::make_numeric_column(data_type(type_to_id<size_type>()), input.size());
+  auto output_mutable_view = output->mutable_view();
+  auto input_device_view   = column_device_view::create(input);
+
+  thrust::transform(thrust::device,
+                    input_device_view->pair_begin<T, InputIsNullable>(),
+                    input_device_view->pair_end<T, InputIsNullable>(),
+                    output_mutable_view.begin<size_type>(),
+                    // Must specify const T as the template type because the column
+                    // views provided on the edges will always return const types. The
+                    // template arguments to `datq` need not specify const since that
+                    // const is added as part of the signature.
+                    bin_finder<const T, LeftComparator, RightComparator>(
```

Comment:
Instead of `const T` here, couldn't it just be `T` and then internally `device_span<T const>`?

Reply:
Yes, it could. I just chose the option that required fewer `const`s :) I'll switch it.